### ISSUE TITLE
On Premise Instance fails to install on Ubuntu 16.04

### ISSUE THREAD

**Issue by chriskinsman** [aws/aws-codedeploy-agent#82]
Running command:

aws deploy install --override-config --config-file codedeploy.onpremises.yml --region us-west-2

Get:

Ign:1 http://cran.fhcrc.org/bin/linux/ubuntu trusty/ InRelease
Hit:2 http://cran.fhcrc.org/bin/linux/ubuntu trusty/ Release  
Hit:4 http://us.archive.ubuntu.com/ubuntu xenial InRelease  
Get:5 http://us.archive.ubuntu.com/ubuntu xenial-updates InRelease [95.7 kB]  
Get:6 http://security.ubuntu.com/ubuntu xenial-security InRelease [94.5 kB]  
Hit:7 http://ppa.launchpad.net/brightbox/ruby-ng/ubuntu xenial InRelease  
Ign:8 http://nginx.org/packages/ubuntu trusty InRelease  
Hit:9 http://us.archive.ubuntu.com/ubuntu xenial-backports InRelease  
Hit:10 http://nginx.org/packages/ubuntu trusty Release  
Fetched 190 kB in 0s (260 kB/s)  
Reading package lists... Done
Reading package lists... Done
Building dependency tree  
Reading state information... Done
ruby2.0 is already the newest version (2.0.0.648-2bbox1~xenial2).
0 upgraded, 0 newly installed, 0 to remove and 5 not upgraded.
Installing the AWS CodeDeploy Agent... ERROR
Failed to stop the AWS CodeDeploy Agent:
Failed to stop codedeploy-agent.service: Unit codedeploy-agent.service not loaded.

Install the AWS CodeDeploy Agent on the on-premises instance by following the instructions in "Configure Existing On-Premises Instances by Using AWS CodeDeploy" in the AWS CodeDeploy User Guide.

Can't find a log file and been trying to figure out why it may not be loading...

**Comment by jdextraze** [comment]
Not sure if this is related, but I had an issue also today with code deploy install... it was looking for ruby-2.0.x specifically

I am doing my install with
```
apt-get -y install ruby2.3
aws s3 cp s3://aws-codedeploy-us-west-2/latest/install . --region us-west-2
chmod +x ./install
./install auto
```

It seems to be related to this commit even if it's old: https://github.com/aws/aws-codedeploy-agent/commit/74e87fd0508fdf3690130247b32d7a40d1bc8441

**Comment by ambareesha** [comment]
jdextraz@, you are facing this issue because we haven't yet updated the new install script in our s3 bucket. We will be doing that with our upcoming host agent release. So, even though the code on github looks like it should run on ruby 2.x, the installers and the install script still enforce the 2.0 dependency.

You can workaround this problem by using the install script on github until we update the version in our s3 bucket:
wget https://raw.githubusercontent.com/aws/aws-codedeploy-agent/master/bin/install
chmod +x ./install
sudo ./install auto

**Comment by jdextraze** [comment]
@ambareesha then why does one of my server who got that script from s3://aws-codedeploy-us-west-2/latest/install on October 1st have the latest change? We didn't change anything in our launch configuration script and it didn't work yesterday, because it was looking for ruby 2.0.x

**Comment by jdextraze** [comment]
Re-downloaded script on that server, you can see the difference in time from the bucket
```
-rw-r--r--   1 root root   13359 Oct 13 19:26 install
-rwxr-xr-x   1 root root   13377 Aug 10 17:35 install.bak
```
Also that server, without any changes, doesn't work anymore within our code deploy since yesterday.

**Comment by jdextraze** [comment]
@chriskinsman sorry to have hijacked your ticket, hope this is related to your issue.

**Comment by ambareesha** [comment]
@jdextraze, did you install ruby2.0 or ruby2.3 on your system?
@chriskinsman, which version of ubuntu are you running? I am checking if I can repro the issue

**Comment by ambareesha** [comment]
@chriskinsman, ignore my previous question. I see that it's Ubuntu 16.04

**Comment by jdextraze** [comment]
@ambareesha we got it working by installing both ruby 2.0 and 2.3. But I am pretty sure the script in the bucket changed yesterday and that's what stopped our auto scaling group from working. It's now working so I wont waste anymore time on this.

**Comment by ambareesha** [comment]
I am able to repro your issue and am working towards finding the root cause.

In the meantime, I find that I am able to install the agent by downloading and running the the install script (http://docs.aws.amazon.com/codedeploy/latest/userguide/how-to-run-agent-install.html#how-to-run-agent-install-ubuntu). Make sure you copy the on-premises configuration file to /etc/codedeploy-agent/conf as described in http://docs.aws.amazon.com/codedeploy/latest/userguide/how-to-configure-on-premises-host.html#how-to-configure-on-premises-host-manual-step-4

**Comment by ambareesha** [comment]
@chriskinsman , the above note is for your issue.

**Comment by ambareesha** [comment]
Figured out the root cause. This is a bug on our side that should affect only on-premise host-agent installations performed via the AWS CLI command "aws deploy install"

**Comment by alexfd** [comment]
Is ruby2.0 still required on 16.04?
I get the following error with v1.0-1.1067:
`2017-03-01 00:55:04 ERROR [codedeploy-agent(30017)]: booting child: error during start or run: TypeError - no implicit conversion of Fixnum into String - /opt/codedeploy-agent/lib/instance_agent/plugins/codedeploy/onpremise_config.rb:30:in `[]='
/opt/codedeploy-agent/lib/instance_agent/plugins/codedeploy/onpremise_config.rb:30:in `configure'
/opt/codedeploy-agent/lib/instance_agent/plugins/codedeploy/command_poller.rb:20:in `initialize'
/opt/codedeploy-agent/lib/instance_agent/agent/base.rb:10:in `new'
/opt/codedeploy-agent/lib/instance_agent/agent/base.rb:10:in `runner'
/opt/codedeploy-agent/lib/instance_agent/runner/child.rb:31:in `block in prepare_run'
/opt/codedeploy-agent/lib/instance_agent/runner/child.rb:55:in `with_error_handling'
/opt/codedeploy-agent/lib/instance_agent/runner/child.rb:30:in `prepare_run'
/opt/codedeploy-agent/vendor/gems/process_manager-0.0.13/lib/process_manager/child.rb:64:in `block in prepare_run_with_error_handling'
/opt/codedeploy-agent/lib/instance_agent/runner/child.rb:55:in `with_error_handling'
/opt/codedeploy-agent/vendor/gems/process_manager-0.0.13/lib/process_manager/child.rb:63:in `prepare_run_with_error_handling'
/opt/codedeploy-agent/vendor/gems/process_manager-0.0.13/lib/process_manager/child.rb:20:in `start'
/opt/codedeploy-agent/vendor/gems/process_manager-0.0.13/lib/process_manager/master.rb:206:in `block in spawn_child'
/opt/codedeploy-agent/vendor/gems/process_manager-0.0.13/lib/process_manager/master.rb:204:in `fork'
/opt/codedeploy-agent/vendor/gems/process_manager-0.0.13/lib/process_manager/master.rb:204:in `spawn_child'
/opt/codedeploy-agent/vendor/gems/process_manager-0.0.13/lib/process_manager/master.rb:283:in `block (2 levels) in replace_terminated_children'
/opt/codedeploy-agent/vendor/gems/logging-1.8.1/lib/logging/diagnostic_context.rb:323:in `block in create_with_logging_context'
2017-03-01 00:55:04 ERROR [codedeploy-agent(30017)]: booting child: error during start or run: SystemExit - exit - /opt/codedeploy-agent/lib/instance_agent/runner/child.rb:67:in `exit'
/opt/codedeploy-agent/lib/instance_agent/runner/child.rb:67:in `rescue in with_error_handling'
/opt/codedeploy-agent/lib/instance_agent/runner/child.rb:55:in `with_error_handling'
/opt/codedeploy-agent/lib/instance_agent/runner/child.rb:30:in `prepare_run'
/opt/codedeploy-agent/vendor/gems/process_manager-0.0.13/lib/process_manager/child.rb:64:in `block in prepare_run_with_error_handling'
/opt/codedeploy-agent/lib/instance_agent/runner/child.rb:55:in `with_error_handling'
/opt/codedeploy-agent/vendor/gems/process_manager-0.0.13/lib/process_manager/child.rb:63:in `prepare_run_with_error_handling'
/opt/codedeploy-agent/vendor/gems/process_manager-0.0.13/lib/process_manager/child.rb:20:in `start'
/opt/codedeploy-agent/vendor/gems/process_manager-0.0.13/lib/process_manager/master.rb:206:in `block in spawn_child'
/opt/codedeploy-agent/vendor/gems/process_manager-0.0.13/lib/process_manager/master.rb:204:in `fork'
/opt/codedeploy-agent/vendor/gems/process_manager-0.0.13/lib/process_manager/master.rb:204:in `spawn_child'
/opt/codedeploy-agent/vendor/gems/process_manager-0.0.13/lib/process_manager/master.rb:283:in `block (2 levels) in replace_terminated_children'
/opt/codedeploy-agent/vendor/gems/logging-1.8.1/lib/logging/diagnostic_context.rb:323:in `block in create_with_logging_context'
2017-03-01 00:55:04 INFO  [codedeploy-agent(29976)]: master 29976: Received CHLD - cleaning dead child process
2017-03-01 00:55:04 INFO  [codedeploy-agent(29976)]: master 29976: been told to replace child 30017
2017-03-01 00:55:04 INFO  [codedeploy-agent(29976)]: master 29976: not enough child processes running - missing at least 1 - respawning`

**Comment by feverLu** [comment]
The new revision of host agent that fix the strict dependency on ruby 2.0 has been release. This problem should already be solved.
